Remove some placeholders used in UI tests #4585

wkurniawan07 opened the pull request:
Fixes #4585

wkurniawan07 commented:
Dr @damithc I believe the last commit here is what you said is good to be on another branch/PR. No doubt it will work on PR, but it has to be done after removing the comment date placeholder or else merge conflicts will happen (which is why I combined in the first place, and also because the change in the last commit is quite minor). I can either make another PR with only the last commit:
1. On top of this PR (instead of master) and merge everything together
2. On top of master after this is merged (without the last commit)

damithc commented:
Option 2 preferred. We don't have a workflow for a PR on top of a PR. 
The reason I insist on a split is because multiple changes in the same PR affects undo-ability of each fix and weakens the code review (more changes means harder to spot problems).

wkurniawan07 commented:
Removed the last commit.

thyageshm commented on src/test/java/teammates/test/cases/ui/browsertests/AllAccessControlUiTests.java:
```diff
@@ -72,7 +73,8 @@ public static void classSetup() {
         
-        currentPage.logout().verifyHtml("/login.html");
+        currentPage.logout();
+        AppPage.getNewPageInstance(browser, IndexPage.class);
```
nice that you have standardised this :+1:

thyageshm commented on src/test/java/teammates/test/pageobjects/IndexPage.java:
```diff
@@ -0,0 +1,15 @@
+    @Override
+    protected boolean containsExpectedPageContents() {
+        String pageSource = getPageSource();
+        return pageSource.contains("Student Login") && pageSource.contains("Instructor Login");
```
I think this might be a very precarious test condition. Isn't there any other string indicative of the index page? Worst case scenario, we can plant some unique string that can be used to uniquely identify the page.

thyageshm replied:
consider looking for ```<span class="submissionsNumber">```

wkurniawan07 replied:
Changed to looking for this (pretty much because all other `containsExpectedPageContents()` also look for the main heading):
![image](https://cloud.githubusercontent.com/assets/7261051/11557169/5bc24c10-99e6-11e5-847a-6ba148c0f923.png)

thyageshm commented:
Everything looks good. Just consider the suggestion on identifying the index.html page

thyageshm commented on src/test/java/teammates/test/pageobjects/AppPage.java:
```diff
@@ -901,12 +897,8 @@ private static String replaceUnpredictableValuesWithPlaceholders(String content)
-                // dynamic feedback submission numbers
-                .replaceAll("(?s)<span class=\"submissionsNumber\".*?</span>", "<span class=\"submissionsNumber\" id=\"submissionsNumber\">\\${submissions\\.number}</span>")
+                // now (date, time) e.g. [Thu, 07 May 2015, 07:52 PM] or [Thu, 07 May 2015, 07:52:13 UTC]
+                .replaceAll(new SimpleDateFormat("EEE, dd MMM yyyy, ").format(now) + "[0-9]{2}:[0-9]{2}( [AP]M|:[0-9]{2} UTC)", "\\${datetime\\.now}")
```
Hey just take note that this is alright only if we are going to standardize the date format soon. We dont want to combine different formats into one placeholder as it reduces our ability to check the different formats.

thyageshm replied:
@wkurniawan07 if we are going to standardize, please create an issue for it before we close this pr

wkurniawan07 replied:
I hardly see the point of checking for different formats; they're essentially the same thing, unless they play for a functionality beyond displaying (which they don't). That aside, I've actually standardized it if you managed to catch it before I took down the commit.

wkurniawan07 replied:
Check [this](https://github.com/TEAMMATES/repo/compare/4585-reduce-placeholders...4590-datetime-comments?w=1) if you want to see the what I'm going to include in that issue. Perhaps you can even review it before the PR is made.

thyageshm replied:
yes, I do know that you had standardized it before the prof asked you to separate it out. I surfaced this to address a different point.

Although, the difference might have simply been for visual/display purpose, it still can be seen as a "feature" to be tested. By combining it we wont know which format was being used where and whether the intended format came in the intended place.

wkurniawan07 replied:
Yes I understand what you're trying to say, but let me remind you of a similar case: the standardized truncated accounts. We agreed to suppress truncated accounts back to its original when we could have used `${test.truncated.account}`s, because what's the point? More placeholders means more `.replace(...)` which makes the comparison slower, and the only advantage is we know that the accounts are truncated in places where they should be truncated, which is easily achieved by hard-coding _just one_ test case to confirm it.

This standardization of placeholders is no different. Only that there is not just one hard-coded comment dates in test files, but hundreds.

And I do realize that combining the regexes (`( [AP]M|:[0-9]{2} UTC)`) is not that much faster compared to the two separate replaces, which is why I attempted to standardize also the date/time format. My reason for removing the placeholders should have been obvious: for performance purpose.

damithc replied:
my 2 cents: In principle the display format should be considered a feature and combined regex weakens testing. We want the test to fail if the time is not in the format intended for that page. The same reasoning applies to truncation, but in that case weakening the test can be justified because the truncation is not predictable (it depends on the developer's choice of test accounts), and the truncation is tested at a lower layer.
Anyway, this is no longer relevant (i.e. we can remove the combined regex) after the time is standardized right?

wkurniawan07 replied:
1. Just because we standardize the date/time format, it does _not_ necessarily translate to the same placeholder used (see how our regkey, question ID, ... have the same regex?). It will be breaking abstraction, something like "it just happened that the same date/time format is used in both places". Then why do I standardize the placeholder? Like I mentioned in the issue thread, `${comment.date}` makes little to no sense and it really refers to date/time now formatted for comments.
2. In both cases of truncated accounts and different date/time formats, the placeholders are _not_ the only place where the different formats are tested. In the former case we have a test case with hard-coded long account to test whether the truncation occurs. In the latter case, we have _hundreds_ of instances of date/time format of comments. If the placeholder fails its job, the hard-coded instances will cover it.

Then why is standardization of date/time format involved here? Once again, standardization is never my primary objective; optimization is, and it is achieved by reducing the number of `.replace(...)`. `${comment.date}` happens to be one that can be removed. I can either do this:
```
.replaceAll(new SimpleDateFormat("EEE, dd MMM yyyy, ").format(now) + "[0-9]{2}:[0-9]{2}:[0-9]{2} UTC", "\\${datetime\\.now}")
.replaceAll(new SimpleDateFormat("EEE, dd MMM yyyy, ").format(now) + "[0-9]{2}:[0-9]{2} [AP]M", "\\${datetime\\.now}")
```
which has no performance boost, or this (currently):
```
.replaceAll(new SimpleDateFormat("EEE, dd MMM yyyy, ").format(now) + "[0-9]{2}:[0-9]{2}( [AP]M|:[0-9]{2} UTC)", "\\${datetime\\.now}")
```
which might have little to no performance boost I'm not sure, or this (what I'm aiming for):
```
.replaceAll(new SimpleDateFormat("EEE, dd MMM yyyy, ").format(now) + "[0-9]{2}:[0-9]{2} [AP]M( UTC)?", "\\${datetime\\.now}")
```
which has obvious performance boost.

damithc replied:
May be my understanding of this situation is incomplete from the beginning. Let's take the truncation example. Let's say there are 8 places in the UI where we use truncated Google ID and 10 places where we use the full ID. Now let's say a certain PR incorrectly changed a jsp page such that one of those 8 places is now using full ID but no html files used for testing was changed by the PR. Will that result in a test failure because the test expects the truncated ID but the actual html has the full ID?
If the answer is 'no', that is, a dev can change the jsp page without changing the html page and still pass the test cases, then our tests are not as strong as they can be. That is, the tests accepts either full ID or truncated ID and doesn't care which one is given.
My understanding is, in our current code, a dev can change a jsp page from 'truncated' to 'full' (or vice versa) without changing html pages, and still pass the tests. Am I right?

Similarly, let's assume we have two date formats. (Although you have now standardized it to one format, it is quite likely we will have multiple formats in future. e.g. a short format and a long format.)
If a dev can change a jsp from one format to the other, and still pass the tests (without changing any html files), then the tests are weak. If the tests are strong, any change to UI should be accompanied by a corresponding changes to test html files.

Now, if we are weakening our tests in the above manner, it should be weighed against the performance (or any other) gains we are getting in return.

wkurniawan07 replied:
No, that is wrong. It might be true for certain tests, but as long as the full test suite is concerned, _all_ is covered. This is what I've been trying to say from the beginning: we have hard-coded test cases that will cover such cases to prevent flexibility even in things as small as display formats.
By using same placeholder, yes I am weakening the test _which uses the test accounts or date/time now_, but there are hundreds of hard-coded test cases that cover it and the weakening becomes a very small price to pay.

This calls for concrete examples:
1. [This file](https://github.com/TEAMMATES/repo/blob/master/src/test/resources/pages/InstructorCourseStudentDetailsRegistered.html). The google ID in question is `CCSDetailsUiT.instr.veryLongGoogleId`, long enough to necessitate truncation. And in that test file, _it is very specific_ where `CCSDetailsUiT.instr.veryLongGoogleId` and `CCSDetailsUiT.instr....` have to appear. Had someone change where truncation occurs without changing this test file, dev green is a big no.
2. [This file](https://github.com/TEAMMATES/repo/blob/master/src/test/resources/pages/studentCourseDetailsWithTeammatesHTML.html). Same case as above, the google ID is `SCDetailsUiT.alice.veryLongGoogleId`.
3. Now for the comment date/time format: [see how much changes are made in the HTML files just by changing the date/time format](https://github.com/TEAMMATES/repo/compare/4585-reduce-placeholders...4590-datetime-comments?w=1). The message is the same: _it is very specific_ where the date/time format for comments appear, and we don't need another placeholder just to make sure of that.

damithc replied:
OK, I think I understand the situation a bit better now. So, let me restate my query.
Tests can verify truncation at three levels.
1. Does the truncation function work? This is basically a unit test of the truncation function. 
2. Does the truncation function work correctly when used in the UI? This needs only one integration test.
3. Is the truncation function used in **all** places where it is supposed to be used?
I assume 1 and 2 are verified by our test suite. So the point of contention is about 3. Does our test suite cover 3? i.e. Is there **at least one place** where we can change the jsp file without changing the html file and sill pass the test?

wkurniawan07 replied:
3 is not entirely true. There can be some places where truncation can be applied and yet the test will still pass, but this is because we didn't provide hard-coded test cases (i.e those with long google IDs), not because of the standardized placeholders. For the places with standardized placeholders, I have guarded them with enough hard-coded test cases.

Anyway, this has been a good discussion, but has gone well off-tangent w.r.t. the actual issue: the date/time format for comments. Date/time format for comments easily score a "yes" for all the 3 queries above.

damithc replied:
Nothing more from me regarding this issue.

thyageshm commented:
Once the new issue is created, this pr can be merged

damithc commented on src/test/resources/data/AutomatedSessionRemindersTest.json:
```diff
@@ -22,7 +22,7 @@
       "courseId": "AutSessRem.course",
       "name": "Test Ins for Aut Sessions Reminder",
-      "email": "${support.email}",
+      "email": "${test.instructor}@gmail.com",
```
This change doesn't look right. Either the original was wrong or the new one is wrong. From what I remember, we had a concept of 'support email address'. This is the email address that will be given in the TEAMMATES emails as the contact point, and probably as the 'reply to' address. Did we get rid of that concept at some point? If we did, we shouldn't have. If we didn't, then why was support email was used here originally and now it is being changed to test instructor's email?

damithc replied:
It is also not clear how this change is related to this PR.

wkurniawan07 replied:
Putting the explanation here just in case anyone refers back to this PR in the future.

I am NOT trying to make the concept of support email address obsolete here, because the support email address is of course of great importance. What I am intending to make obsolete is its placeholder, i.e the `${support.email}`.

As far as this test is concerned, neither the original nor the new one is wrong, because we only want to check our chosen email box(es) to see if the session reminder emails are sent, but the original choice required us to have `${support.email}`, and that placeholder comes bundled with `.replace("${support.email}", Config.SUPPORT_EMAIL)` which is going to be applied to **ALL** UI test comparison, and that's highly redundant especially when _only one_ test uses it (I have verified this).

Which leads to the next question: why did we specifically use support email here? I did not manage to find an answer to that, because replacing it with any of the test accounts does not in any way affect the strength of the test, it only changes which email box(es) we should look the reminder emails in. With that in mind, I am confident that the placeholder `${support.email}` can be removed once and for all and here we go.

damithc replied:
I think now I recall why the `support.email` is used here. It's because during live testing emails are actually sent out which end up in my email box. So when I do a live testing, I expect to see a certain set of emails in my email box. If I don't see them, I know there is a breakdown in our email sending. So the original one was correct, but it was not documented properly. This manual check while live testing is important because email sending is skipped in dev server testing.

wkurniawan07 replied:
Yes I know about that, it was documented in `AutomatedSessionRemindersTest.java`, and I also did the testing on staging server myself. However the point here is why _specifically_ support email address while other email accounts work just fine (I have also tested this).

damithc replied:
That's because of the way we do live testing. 
1. emails sent to the support email is supposed to come to my inbox. This test ensures that it indeed does happen. 
2. If I use the actual support email address for a different test account, (e.g. for test.instructor), then I get spammed too much during testing.

wkurniawan07 replied:
I'm not saying that the same email account should be used for `${test.instructor}`. You can simply check the `${test.instructor}` email box, whatever account used for it, to see if the emails arrive. I'm not sure if checking more than one email boxes (if it's not already done that way) is more troublesome than having the test slowed down because of the extra `.replace(...)`.

damithc replied:
Pros of this approach:
Simplifies the tests a bit. That's always a good thing.
Speeds up the tests. Any idea by how much? I'm guessing a few seconds in total, but I could be wrong.

Cons of this approach:
1. Right now I don't have to check any additional email boxes because in the live server support.email is set up to reach my own inbox. If you change it like this, I have to check additional email boxes. That's what I'm trying to avoid. 
2. The original approach automatically confirms that the support.email is correctly directed to my email. There is a redirection involved in the path (from teammates@comp.... account to my gmail account) so a periodic validation of the path is useful because the path can break without me realizing it. e.g. due to spam filter added by SoC, expiry of unix account etc.
3. If we reuse an account here that is used in other tests (you are reusing two existing test accounts in this fix), it will be hard to detect if these specific emails because not only I have to check two additional accounts, I have to detect these specific emails among many other emails triggered by other tests.  

So the cons may not affect other devs, but they do make my life harder.

wkurniawan07 replied:
I see, so the trade-off is not as simple as I thought before.
Let me give this a thought for a while, I should be able to come up with a solution that addresses the cons as well.

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/AutomatedSessionRemindersTest.java:
```diff
@@ -28,6 +29,20 @@ public static void classSetup() throws Exception {
+        testData.feedbackSessions.get("closingSession").creatorEmail = Config.SUPPORT_EMAIL;
+        testData.feedbackSessions.get("openingSession").creatorEmail = Config.SUPPORT_EMAIL;
+        testData.feedbackSessions.get("publishedSession").creatorEmail = Config.SUPPORT_EMAIL;
+        testData.feedbackQuestions.get("question").creatorEmail = Config.SUPPORT_EMAIL;
```
Dr @damithc I can't say that this is an elegant solution, but this is the only way I can think of that both achieves what I want to achieve and also doesn't make your life harder.

damithc replied:
I think this is fine. Doing this in the java file is in fact is better than doing it in the json file.
In the json file, may be set the emails to some obvious value? e.g. `this_value_will_be_overridden`

wkurniawan07 replied:
Consider it done :+1:

Anyway, if this approach is better, we might as well move all other injections to the respective java files as well, there's not a lot of them. Advantages:

1. We can explain in the java files why the particular accounts are used.
2. We can remove the injection process to the json data files, which takes time and is applied to all regardless of presence of placeholders. This is again another optimization no matter how slight.

Of course, this is another issue.

damithc replied:
No objections.

thyageshm commented:
Hi, sorry I came home and internet connection here has been intermittent. I will look through all changes now. I apologise for the delay

thyageshm commented on src/test/java/teammates/test/cases/ui/browsertests/StudentHomePageUiTest.java:
```diff
@@ -67,14 +67,14 @@ private void testContentAndLogin() throws Exception {
-
-        // This is the full HTML verification for Student Home Page, the rest can all be verifyMainHtml
-        studentHome.verifyHtml("/studentHomeHTMLEmpty.html");
+        studentHome.verifyHtmlMainContent("/studentHomeHTMLEmpty.html");
```
we can consider providing reason here as to why this one is MainContent and the full html check comes later. Even if we don't spell the reason out in full, we should at least caution the developer from changing this. Also provide a quick reference to this pr/issue.

thyageshm commented on src/test/resources/data/AutomatedSessionRemindersTest.json:
```diff
@@ -4,7 +4,7 @@
   	  "name": "Test Ins for Aut Sessions Reminder",
   	  "isInstructor": true,
-  	  "email": "${support.email}",
+  	  "email": "${value.to.be.overridden}",
```
I think we can use a more descriptive value that states that the value will be overridden in the test file. This will help a new developer to quickly identify the location where the value is obtained.

Not sure if this is needed. Expected use case is when we generate these values in different places and not just the test file. What do you think?

wkurniawan07 replied:
Makes sense. Will do that in a jiffy.

thyageshm commented:
Two small things and then we are clear

wkurniawan07 commented:
Changes made. Also I removed the `IndexPage`, turns out there's already one made before (`HomePage`), I wonder why it's not used.

thyageshm commented:
Good to go!

wkurniawan07 commented:
@thyageshm hey boss, sorry I made one more set of change (I promise this'll be the last for this PR :p). Just go to the last commit, everything else is the same as before (although I rebased and squashed). This time I removed `${app.url}` by moving the HTML verification to another place with no status message (`${app.url}` only makes a single appearance there). The test is not weakened since the status message is tested via Selenium anyway.

thyageshm commented:
everything looks good. just make sure that there are no other places with app.url
